sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MicroBatchExecution.scala

```diff
@@ -236,14 +257,31 @@ class MicroBatchExecution(
     val hasNewData = {
       awaitProgressLock.lock()
       try {
-        val latestOffsets: Map[Source, Option[Offset]] = uniqueSources.map {
+        val latestOffsets: Map[BaseStreamingSource, Option[Offset]] = uniqueSources.map {
```

Comment:
nit: Can you add more comments in this section to explain what this does? This section is becoming more complicated.

Reply:
I tried, but I'm not quite sure what to add.